taskweaver/code_interpreter/code_generator/code_generator_plugin_only.py

```diff
@@ -0,0 +1,141 @@
+            "compression_prompt_path",
+            os.path.join(
+                os.path.dirname(os.path.abspath(__file__)),
+                "compression_prompt.yaml",
+            ),
+        )
+        self.enable_auto_plugin_selection = self._get_bool("enable_auto_plugin_selection", False)
+        self.auto_plugin_selection_topk = self._get_int("auto_plugin_selection_topk", 3)
+
+
+class CodeGeneratorPluginOnly(Role):
+    @inject
+    def __init__(
+        self,
+        config: CodeGeneratorPluginOnlyConfig,
+        plugin_registry: PluginRegistry,
+        logger: TelemetryLogger,
+        llm_api: LLMApi,
+    ):
+        self.config = config
+        self.logger = logger
+        self.llm_api = llm_api
+
+        self.role_name = self.config.role_name
+
+        self.post_translator = PostTranslator(logger)
+        self.prompt_data = read_yaml(self.config.prompt_file_path)
+        self.plugin_pool = [p for p in plugin_registry.get_list() if p.plugin_only is True]
+        self.instruction_template = self.prompt_data["content"]
+
+        if self.config.enable_auto_plugin_selection:
+            self.plugin_selector = PluginSelector(plugin_registry, self.llm_api)
+            self.plugin_selector.generate_plugin_embeddings()
+            logger.info("Plugin embeddings generated")
+            self.selected_plugin_pool = SelectedPluginPool()
+
+    def select_plugins_for_prompt(
+        self,
+        user_query,
+    ) -> List[PluginEntry]:
+        selected_plugins = self.plugin_selector.plugin_select(
+            user_query,
+            self.config.auto_plugin_selection_topk,
+        )
+        self.selected_plugin_pool.add_selected_plugins(selected_plugins)
+        self.logger.info(f"Selected plugins: {[p.name for p in selected_plugins]}")
+        self.logger.info(f"Selected plugin pool: {[p.name for p in self.selected_plugin_pool.get_plugins()]}")
+
+        return self.selected_plugin_pool.get_plugins()
+
+    def reply(self, memory: Memory, event_handler: callable) -> Post:
+        # extract all rounds from memory
+        rounds = memory.get_role_rounds(
+            role="CodeInterpreter",
+            include_failure_rounds=False,
+        )
+
+        user_query = rounds[-1].user_query
+        if self.config.enable_auto_plugin_selection:
+            self.plugin_pool = self.select_plugins_for_prompt(user_query)
+
+        # obtain the user query from the last round
+        prompt, tools = _compose_prompt(
+            system_instructions=self.instruction_template.format(
+                ROLE_NAME=self.role_name,
+            ),
+            rounds=rounds,
+            plugin_pool=self.plugin_pool,
+        )
+        post = Post.create(message=None, send_from="CodeInterpreter", send_to="Planner")
+
+        llm_response = self.llm_api.chat_completion(
+            messages=prompt,
+            tools=tools,
+            tool_choice="auto",
+            response_format=None,
+            stream=False,
+        )
```

Comment:
do we have to configure the `stream` to False in such context? currently seems in LLM module, the API for streamed and non-streamed is also rather different and not fully align with OpenAI's API either

Reply:
it doesn't make any sense to display the output of the function calling?  this can also simplify the handling of llm response for function calling. the output of the model is actually not json string, but a list of classes.

Reply:
It's not intended to display the intermediate content but try to unify the API call to use stream version (non-stream API call could be regarded as a stream with only 1 chunk). The web API response is still a list of JSON chunks yet just wrapped by the OpenAI Python library. Other LLM implementations won't do the wrapping.